src/crimson/osd/replicated_recovery_backend.h

```diff
@@ -47,16 +47,37 @@ class ReplicatedRecoveryBackend : public RecoveryBackend {
     Ref<MOSDPGRecoveryDelete> m);
   interruptible_future<> handle_recovery_delete_reply(
     Ref<MOSDPGRecoveryDeleteReply> m);
+  interruptible_future<PushOp> prep_push_to_replica(
+    const hobject_t& soid,
+    eversion_t need,
+    std::pair<pg_shard_t, pg_missing_t> pg_shard);
+  void calc_clone_subsets(
+    SnapSet& snapset, const hobject_t& soid,
+    const pg_missing_t& missing,
+    const hobject_t &last_backfill,
+    interval_set<uint64_t>& data_subset,
+    std::map<hobject_t, interval_set<uint64_t>>& clone_subsets);
+  void calc_head_subsets(
+    crimson::osd::ObjectContextRef obc,
+    SnapSet& snapset,
+    const hobject_t& head,
+    const pg_missing_t& missing,
+    const hobject_t &last_backfill,
+    interval_set<uint64_t>& data_subset,
+    std::map<hobject_t, interval_set<uint64_t>>& clone_subsets);
```

Comment:
These two functions encapsulate a mildly complicated amount of logic.  Moreover, their implementations don't actually rely on any members of ReplicatedRecoveryBackend.  Thus, it's worth arranging things to allow us to unit test them in isolation:
- calc_head_subsets doesn't seem to need a full ObjectContextRef, it only seems to access obc->obs.oi.size
- Move to a common header (crimson/osd/object_metadata_helpers.h/cc)
- Return clone_subsets and data_subset in a struct by value.
- Add at least a single unit test for each.